### DISCUSSION TITLE
Implicit Configuration replacement doubts

### DISCUSSION THREAD

**Discussion by ShadyD45** [vladimir-bukhtoyarov/bucket4j#310] [Q&A] [answered]
Hello @vladimir-bukhtoyarov !
I'm little confused about implicit Configuration replacement. could you please clarify some of the below doubts?

1) when is the bucket config replaced? I found that Java doc states `When implicit configuration replacement is activated, bucket will check that version of configuration in the storage >= than provided desiredConfigurationVersion` but in [8.1.0 reference](https://bucket4j.com/8.1.0/toc.html#implicit-configuration-replacement) docs it says `when bucket detects that persisted configuration version is less that provided through builder API`. Both mean different conditions. Please correct me if I've understood it wrong.

2) Do we have to take care of incrementing or decrementing (based on above condition) the `desiredConfigurationVersion` to meet the criteria for replacement? If yes could you please suggest how would you do it? Because to increment or decrement we'll have to check if the config is changed but then if we have to do that it defeats the purpose of implicit replacement right?

It would be great if we could add a few more example scenarios and explanation for implicit configuration replacement in the doc.
Thankyou!!

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @ShadyD45 

1. Config replacements happens if persisted version < provided version. I have rewrote documentation to be more clearr.

2. Implicit replacement initially was designed for cases where configuration is stored inside config files. For example imaging that you have some bucket with configuration that is stored in a properties file like this:
```
...
bucket.capacity=100
bucket.refill.period.seconds=60
bucket.refill.tokens=50
bucket.version=1
...
```
In case of in new version of application you want to change bucket configuration, you need just to update properties file:
```
...
bucket.capacity=150
bucket.refill.period.seconds=60
bucket.refill.tokens=50
bucket.version=2
...
```
It can be another use cases for implicit configuration use cases but I can not image something.

**Reply by ShadyD45** [reply]
Thankyou for your response @vladimir-bukhtoyarov 
In my scenario config don't exactly come from a properties file but from your explanation what I understand is that I have to maintain the version of each bucket somewhere and update the version whenever the config is updated. And pass that version to `withImplicitConfigurationReplacement()` whenever i fetch the bucket and it'll give me the updated config if replacement condition is satisfied and also store the new config in my distributed storage. Am I right?

**Reply by vladimir-bukhtoyarov** [reply]
In such complex case, you should not use implicit config replacement. It was designed to make code more simple, not to make code more harder. [Explicit replace](https://bucket4j.com/8.1.0/toc.html#configuration-replacement) looks to be enough, just call it each time when you change configuration inside configs storage.

**Reply by ShadyD45** [reply]
Ok I think I have a clear idea about replacement now. One last question. So I'm fetching bucket from storage with this
 `Bucket bucket = proxyManager.builder().build(key, getBucket());`
 now as you suggested to use explicit replacement. Do I have to just invoke `bucket.replaceConfiguration(newConfig, TokenInheritanceStrategy )` on returned persisted bucket? And it'll persist the new config in storage as well?

**Reply by vladimir-bukhtoyarov** [reply]
Configuration is stored together with another parts of bucket state, so yes, calling ```replaceConfiguration``` is enough.

**Reply by ShadyD45** [reply]
Oh great. Really appreciate your help. Thanks
